File: backend/src/runtime_paths.mjs
import * as path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const DEERFLOW_ROOT = path.resolve(process.env.DEERFLOW_ROOT || path.join(__dirname, '..', '..'));
export const DEERFLOW_RUNTIME_ROOT = path.resolve(
  process.env.DEERFLOW_RUNTIME_ROOT || path.join(DEERFLOW_ROOT, '.deerflow')
);

export function runtimePath(...segments) {
  return path.join(DEERFLOW_RUNTIME_ROOT, ...segments);
}

File: backend/app/runtime_paths.py
"""Project-local DeerFlow runtime paths."""

from __future__ import annotations

import os
from pathlib import Path

DEERFLOW_ROOT = Path(os.environ.get("DEERFLOW_ROOT", Path(__file__).resolve().parents[2])).resolve()
DEERFLOW_RUNTIME_ROOT = Path(os.environ.get("DEERFLOW_RUNTIME_ROOT", DEERFLOW_ROOT / ".deerflow")).resolve()


def runtime_path(*parts: str) -> Path:
    return DEERFLOW_RUNTIME_ROOT.joinpath(*parts)


def upgrade_center_state_dir() -> Path:
    return runtime_path("upgrade-center", "state")
